### PR TITLE
Resume experiment with extra trials from last checkpoint

### PR DESCRIPTION
**Pull request by johnugeorge**
Fixes: #891 
Resuming experiment feature is added.  If user wants to try the same experiment with more number of trials, max trials can be reconfigured to restart the experiment from the last checkpoint. 

Experiment is restarted only when experiment is in succeeded state with max trials reached previously.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/katib/952)

### CODE REVIEW

**Comment by johnugeorge** [comment]
/hold 
Holding till reviews are completed. 
/cc @gaocegege 
/cc @hougangliu

**Comment by johnugeorge** [comment]
@gaocegege  added a test to reconfigure max trails and parallel trial count

File: pkg/util/v1alpha3/katibclient/katib_client.go [hunk 1]
```diff
@@ -123,6 +124,14 @@ func (k *KatibClient) CreateExperiment(experiment *experimentsv1alpha3.Experimen
 	return nil
 }
 
+func (k *KatibClient) UpdateExperiment(experiment *experimentsv1alpha3.Experiment, namespace ...string) error {
```
Comment (hougangliu) [inline pkg/util/v1alpha3/katibclient/katib_client.go]: is this method used for UI? and it seems `namespace` param isn't used

File: pkg/controller.v1alpha3/experiment/experiment_controller.go [hunk 2]
```diff
@@ -186,8 +186,22 @@ func (r *ReconcileExperiment) Reconcile(request reconcile.Request) (reconcile.Re
...
 	}
 
-	if instance.IsCompleted() && !instance.HasRunningTrials() {
-		return reconcile.Result{}, nil
+	if instance.IsCompleted() {
+		// Check if completed instance is restartable
+		// Experiment is restartable only if it is in succeeded state by reaching max trials
+		if util.IsCompletedExperimentRestartable(instance) {
+			// Check if max trials is reconfigured
+			if (instance.Spec.MaxTrialCount != nil) &&
```
Comment (hougangliu) [inline pkg/controller.v1alpha3/experiment/experiment_controller.go]: I think if a user changes `MaxTrialCount` to nil (infinity), here we should also allow to restart it

File: pkg/util/v1alpha3/katibclient/katib_client.go (hunk 1 above)
Comment (johnugeorge) [reply pkg/util/v1alpha3/katibclient/katib_client.go]: I am not sure about it. Shall we do this separately?

File: pkg/controller.v1alpha3/experiment/experiment_controller.go (hunk 2 above)
Comment (johnugeorge) [reply pkg/controller.v1alpha3/experiment/experiment_controller.go]: Updated.

File: pkg/util/v1alpha3/katibclient/katib_client.go (hunk 1 above)
Comment (hougangliu) [reply pkg/util/v1alpha3/katibclient/katib_client.go]: If this method is not useful in your PR topic, I think we'd better remove it.

File: pkg/controller.v1alpha3/experiment/experiment_controller.go [hunk 3]
```diff
@@ -191,8 +191,9 @@ func (r *ReconcileExperiment) Reconcile(request reconcile.Request) (reconcile.Re
 		// Experiment is restartable only if it is in succeeded state by reaching max trials
 		if util.IsCompletedExperimentRestartable(instance) {
 			// Check if max trials is reconfigured
-			if (instance.Spec.MaxTrialCount != nil) &&
-				(*instance.Spec.MaxTrialCount != instance.Status.Trials) {
+			if (instance.Spec.MaxTrialCount != nil &&
+				*instance.Spec.MaxTrialCount != instance.Status.Trials) ||
+				(instance.Spec.MaxTrialCount == nil && instance.Status.Trials != 0) {
```
Comment (hougangliu) [inline pkg/controller.v1alpha3/experiment/experiment_controller.go]: should we consider the case: instance.Spec.MaxTrialCount == 0 and the experiment will be marked completed with instance.Status.Trials == 0; then update instance.Spec.MaxTrialCount to nil?

File: pkg/util/v1alpha3/katibclient/katib_client.go (hunk 1 above)
Comment (johnugeorge) [reply pkg/util/v1alpha3/katibclient/katib_client.go]: No. this method is currently used in resume_e2e_experiment.go script in this PR to update the experiment.

File: pkg/controller.v1alpha3/experiment/experiment_controller.go (hunk 3 above)
Comment (johnugeorge) [reply pkg/controller.v1alpha3/experiment/experiment_controller.go]: i think, this is a invalid case. It doesn't make sense to set MaxTrialCount to be zero if set. We should add a validation for this separately. 
Related: https://github.com/kubeflow/katib/issues/768

**Comment by johnugeorge** [comment]
/hold cancel
as reviews are completed

**Pull request merged**